bpf/tc_http_tp.c

```diff
@@ -0,0 +1,588 @@
+}
+
+[[maybe_unused]] static __always_inline struct ethhdr *eth_header(struct __sk_buff *ctx) {
+    void *data = ctx_data(ctx);
+
+    return (data + sizeof(struct ethhdr) > ctx_data_end(ctx)) ? NULL : data;
+}
+
+static __always_inline struct iphdr *ip_header(struct __sk_buff *ctx) {
+    void *data = ctx_data(ctx);
+
+    data += sizeof(struct ethhdr);
+
+    return (data + sizeof(struct iphdr) > ctx_data_end(ctx)) ? NULL : data;
+}
+
+static __always_inline struct tcphdr *tcp_header(struct __sk_buff *ctx) {
+    void *data = ctx_data(ctx);
+
+    data += sizeof(struct ethhdr) + sizeof(struct iphdr);
+
+    return (data + sizeof(struct tcphdr) > ctx_data_end(ctx)) ? NULL : data;
+}
+
+static __always_inline unsigned char *tcp_payload(struct __sk_buff *ctx) {
+    struct tcphdr *tcp = tcp_header(ctx);
+
+    if (!tcp) {
+        return NULL;
+    }
+
+    unsigned char *payload = (unsigned char *)tcp;
+
+    payload += tcp->doff * 4;
+
+    if ((void *)payload > ctx_data_end(ctx)) {
+        return NULL;
+    }
+
+    return payload;
+}
+
+static __always_inline int update_tcp_csum(struct __sk_buff *ctx) {
+    const struct iphdr *iph = ip_header(ctx);
+
+    if (!iph) {
+        return 0;
+    }
+
+    struct tcphdr *tcp = tcp_header(ctx);
+
+    if (!tcp) {
+        return 0;
+    }
+
+    const unsigned char *b = (const unsigned char *)tcp;
+    const unsigned char *e = (const unsigned char *)ctx_data_end(ctx);
+
+    tcp->check = 0;
+
+    const __u16 tcp_len = (e - b) & MAX_IP_PACKET_SIZE;
+
+    __u32 data_sum = 0;
+
+    __u32 pseudo_header[3];
+    pseudo_header[0] = iph->saddr;
+    pseudo_header[1] = iph->daddr;
+    pseudo_header[2] = bpf_htonl((6 << 16) | tcp_len);
+
+    calculate_datasum_inline(
+        (unsigned char *)&pseudo_header, (unsigned char *)&pseudo_header[3], &data_sum);
+
+    if (tcp_len < MAX_INLINE_LEN) {
+        calculate_datasum_inline((unsigned char *)tcp, e, &data_sum);
+    } else {
+        calculate_datasum((unsigned char *)tcp, e, &data_sum);
+    }
+
+    const __u16 tcp_csum = (__u16)~data_sum;
+
+    tcp = tcp_header(ctx);
+
+    if (!tcp) {
+        return 0;
+    }
+
+    tcp->check = tcp_csum;
+
+    return 1;
+}
+
+static __always_inline long update_ip_csum(struct __sk_buff *ctx) {
+    struct iphdr *iph = ip_header(ctx);
+
+    if (!iph) {
+        return -1;
+    }
+
+    const __u16 new_total_len = bpf_ntohs(iph->tot_len) + EXTEND_SIZE;
+    const __be16 ip_tot_len_old = iph->tot_len;
+    const __be16 ip_tot_len_new = bpf_htons(new_total_len);
+
+    const __u32 ip_csum_off = (sizeof(struct ethhdr) + offsetof(struct iphdr, check)) & 0xff;
+
+    return bpf_l3_csum_replace(
+        ctx, ip_csum_off, ip_tot_len_old, ip_tot_len_new, sizeof(ip_tot_len_new));
+}
+
+static __always_inline __u32 extra_xmited_bytes(__u32 key) {
+    const __u32 *seq = bpf_map_lookup_elem(&seq_offset_map, &key);
+
+    return seq ? *seq : 0;
+}
+
+static __always_inline int is_http_request(struct __sk_buff *ctx) {
```

Comment:
I had to merge the logic from `is_http` here, again, for complexity reasons (in this case, doing without unnecessary branching and storing the `packet_type`).